src/librustc_typeck/check/method/suggest.rs

```diff
@@ -79,65 +115,35 @@ pub fn report_error<'a, 'tcx>(fcx: &FnCtxt<'a, 'tcx>,
-                        _ => {
-                            if let Ok(fn_once_trait_did) =
-                                    cx.lang_items.require(FnOnceTraitLangItem) {
-                                let infcx = fcx.infcx();
-                                infcx.probe(|_| {
-                                    let fn_once_substs =
-                                        Substs::new_trait(vec![infcx.next_ty_var()],
-                                                          Vec::new(),
-                                                          field_ty);
-                                    let trait_ref =
-                                      ty::TraitRef::new(fn_once_trait_did,
-                                                        cx.mk_substs(fn_once_substs));
-                                    let poly_trait_ref = trait_ref.to_poly_trait_ref();
-                                    let obligation = Obligation::misc(span,
-                                                                      fcx.body_id,
-                                                                      poly_trait_ref
-                                                                         .to_predicate());
-                                    let mut selcx = SelectionContext::new(infcx);
-
-                                    if selcx.evaluate_obligation(&obligation) {
-                                        span_stored_function!();
-                                    } else {
-                                        span_did_you_mean!();
-                                    }
-                                });
-                            } else {
-                                span_did_you_mean!();
-                            }
-                        }
+                    if is_fn_ty(&field_ty, &fcx, span) {
+                        err.span_note(span,
+                                      &format!("use `({0}.{1})(...)` if you meant to call \
+                                               the function stored in the `{1}` field",
+                                               expr_string, item_name));
+                    } else {
+                        err.span_note(span, &format!("did you mean to write `{0}.{1}`?",
+                                                     expr_string, item_name));
                     }
                 }
             }
 
+            if is_fn_ty(&rcvr_ty, &fcx, span) {
+                let expr_string = match rcvr_expr {
+                    Some(expr) => match cx.sess.codemap().span_to_snippet(expr.span) {
+                        Ok(expr_string) => expr_string,
+                        _ => "s".into()
+                    },
+                    _ => "s".into()
+                };
+                err.fileline_note(
+                    span,
+                    &format!("method invoked on function type. did you \
```

Comment:
Perhaps the error should be something like

"stdin is a function, perhaps you wish to call it"?

And, _if_ the span_to_snippet call works (don't use "s" as a placeholder snippet), add a `span_suggestion` which suggests you try `{}().{}` or whatever. Don't include suggestions within the notes.